File: summary/ts/3_quick_sort.ts
function quickSort(array: number[]): number[] {
    if (array.length <= 1) {
        return array;
    }

    const pivotIndex = Math.floor(array.length / 2);
    const pivot = array[pivotIndex];

    const left: number[] = [];
    const right: number[] = [];

    for (let i = 0; i < array.length; i++) {
        if (i === pivotIndex) {
            continue;
        }
        if (array[i] < pivot) {
            left.push(array[i]);
        } else {
            right.push(array[i]);
        }
    }

    return [...quickSort(left), pivot, ...quickSort(right)];
}

// テスト
const unsortedArray = [5, 3, 7, 2, 9, 1, 6, 4, 8];
const sortedArray = quickSort(unsortedArray);
console.log("ソート前:", unsortedArray);
console.log("ソート後:", sortedArray);

File: summary/rust/3_quick_sort.rs
fn quicksort<T: PartialOrd + Copy>(arr: &mut [T]) {
   let len = arr.len();
   if len <= 1 {
       return;
   }

   let pivot = arr[len / 2];
   let mut left = 0;
   let mut right = len - 1;

   while left <= right {
       while arr[left] < pivot {
           left += 1;
       }

       while arr[right] > pivot {
           right -= 1;
       }

       if left <= right {
           arr.swap(left, right);
           left += 1;
           right -= 1;
       }
   }

   quicksort(&mut arr[..right + 1]);
   quicksort(&mut arr[left..]);
}

fn main() {
   let mut arr = [5, 3, 2, 8, 1, 7];
   quicksort(&mut arr);
   println!("{:?}", arr); // [1, 2, 3, 5, 7, 8]
}
